### PR TITLE
Better parameter checking

### PR DESCRIPTION
**Pull request by Lehren**
Improve constructor and controller parameter checking.

Check for nulls and types of parameters better.

### CODE REVIEW

File: test/speech-challengeSpec.js [hunk 1]
```diff
@@ -228,12 +268,13 @@ describe('SpeechChallenge API interaction test', () => {
     controller.getSpeechChallenge('4')
       .then(result => {
         const request = window.fetch.calls.mostRecent().args;
-        expect(request[0]).toBe(url);
+        expect(request[0]).toBe(url + '/4');
```
Comment (kvdb) [inline test/speech-challengeSpec.js]: that doesn't seem to be related to parameter checking.

Comment (Lehren) [inline test/speech-challengeSpec.js]: fixed

**Comment by Lehren** [comment]
1.Why does this issue exist

The original parameter checking of `if(name){}` is broken because a boolean could pass that check.

2. Promise usage

What? `getSpeechChallenge` returns a Promise containing a SpeechChallenge as documented.

3. Merge conflicts fixed

**Comment by joggienl** [comment]
For me it wasn't clear that something was broken that needed to be fixed. For future PR's and/or issues I would like to suggest a more verbose and explicit message on what's going on with the PR or issue. So, it's more a note in general.

Rest of my comments can be ignored for now.

Since @kvdb also requested a change I would also like to have his go before we merge this one.

**Comment by kvdb** [comment]
@joggienl no need to wait for such a small change request from me.
